source/common/common/perf_annotation.cc

```diff
@@ -32,6 +32,23 @@ void PerfOperation::record(absl::string_view category, absl::string_view descrip
 // because the constructor is non-trivial due to the contained unordered_map.
 PerfAnnotationContext::PerfAnnotationContext() = default;
 
+void PerfAnnotationContext::begin(absl::string_view category, absl::string_view description) {
+  const CategoryDescription key = {std::string(category), std::string(description)};
+  RELEASE_ASSERT(!timestamps_map_.contains(key),
+                 "double perf measurement beginning is disallowed!");
+  timestamps_map_.insert({key, currentTime()});
+}
+
+void PerfAnnotationContext::end(absl::string_view category, absl::string_view description) {
+  const CategoryDescription key = {std::string(category), std::string(description)};
+  const MonotonicTime end_time = currentTime();
+  const auto result = timestamps_map_.extract(key);
+  RELEASE_ASSERT(!result.empty(), "double perf measurement ending is disallowed!");
```

Comment:
might as well include the category & description in this message and the RELEASE_ASSERT above.